### PR TITLE
fix(embeddings): distribute worker requests across replicas (ENG-2561)

### PR DESCRIPTION
**Pull request by BhagyaAmarasinghe**
## What does this PR do?

- Tracks the EU embedding throughput regression under `ENG-2561`.
- Adds the default-off `EMBEDDING_HTTP_DISABLE_KEEP_ALIVES` worker setting.
- Uses a cloned HTTP transport with keep-alive disabled for configured OpenAI-compatible embedding clients, allowing Kubernetes to rebalance each background batch connection across TEI endpoints.
- Leaves Hub API semantic-search traffic and every existing deployment unchanged unless the setting is explicitly enabled on the worker.
- Logs the effective transport setting whenever an embedding worker is configured, including when batching is disabled.
- Propagates the setting through the producer-only `backfill-embeddings` command for configuration consistency; embedding requests are still executed by `hub-worker`.

## Why this changes the staging throughput path

- The failed staging gate ran against six Ready TEI replicas but successful provider requests reached only two pods, while the worker was capped at three in-flight batches; the result was 500 embeddings in 382 seconds (`1.31 embeddings/s`).
- The companion [Formbricks chart PR #9001](https://github.com/formbricks/formbricks/pull/9001) exposes the setting only on background worker/backfill workloads and renders the tested `48 / 8 / 100 / 12` worker tuning.
- The guarded 500-record staging rerun remains the acceptance gate: at least `5.0 embeddings/s`, all six pods used within the distribution bounds, queue drained, and foreground search healthy.

## How should this be tested?

- `make build`
- `make build-backfill-embeddings`
- `make test-unit`
- `go test -race ./internal/config ./internal/openai ./internal/service`
- `make GOLANGCI_LINT="$(go env GOPATH)/bin/golangci-lint" lint`
- Confirm `TestCreateEmbedding_DisableKeepAlivesControlsConnectionReuse` observes one connection by default and two connections for two requests when enabled.
- Confirm `TestBatchingEmbeddingClientLimitsConcurrentProviderRequests` reaches but never exceeds 12 provider requests.

### Required

- [x] Filled out the "How to test" section in this PR
- [x] Read [Repository Guidelines](https://github.com/formbricks/hub/blob/main/AGENTS.md)
- [x] Self-reviewed my own code
- [x] Commented on my code in hard-to-understand bits
- [x] Ran `make build`
- [ ] Ran `make tests` (integration tests require pgvector; this transport-only change is covered by unit and race tests above)
- [x] Ran `make fmt` and `make lint`; no new warnings
- [x] Removed debug prints / temporary logging
- [x] Started from the latest `origin/main`
- [x] No database schema change

### Appreciated

- [x] No API or OpenAPI change
- [x] Updated `.env.example` for the optional setting
- [ ] `make tests-coverage` was not run; focused connection and concurrency regression tests cover the new behavior

### CODE REVIEW

File: internal/service/embedding_batcher_test.go [hunk 1]
```diff
@@ -291,6 +292,51 @@ func TestBatchingEmbeddingClientShutdownFlushesAndCloses(t *testing.T) {
...
+	defer c.active.Add(-1)
+
+	for {
+		maximum := c.maximum.Load()
+		if active <= maximum || c.maximum.CompareAndSwap(maximum, active) {
+			break
+		}
+	}
+
+	c.started <- struct{}{}
```
[the remark this answers is not in the document]
Comment (BhagyaAmarasinghe) [reply internal/service/embedding_batcher_test.go]: Addressed in 2f6e993: the test provider notification is now non-blocking, so teardown cannot deadlock after the assertion stops draining the channel. `go test -race ./internal/service` and lint pass.

File: .env.example [hunk 2]
```diff
@@ -124,6 +124,7 @@ WEBHOOK_MAX_COUNT=500
 # EMBEDDING_BATCH_SIZE=1             (document micro-batch size; 1 disables batching; provider must support batches)
 # EMBEDDING_BATCH_MAX_WAIT_MS=25     (maximum time to collect a partial batch; default 25ms)
 # EMBEDDING_BATCH_MAX_IN_FLIGHT=1    (maximum concurrent provider batch requests; default 1)
+# EMBEDDING_HTTP_DISABLE_KEEP_ALIVES=false (open a new provider connection per request; useful for Kubernetes service distribution)
```
[the remark this answers is not in the document]
Comment (BhagyaAmarasinghe) [reply .env.example]: Addressed in 2f6e993: the env example now states that the setting applies to OpenAI-compatible provider requests.

**Review by xernobyl — COMMENTED** [review]
Nice, focused change and the transport test is the good kind — I reverted the 5-line block in `NewClient` and `opens_a_connection_per_request_when_disabled` went red (`expected 2, actual 1`), so it's really guarding the behaviour. 👍

I set up a fake 3-pod TEI service (each accepted TCP connection assigned round-robin to a "pod", mimicking kube-proxy) and ran the real worker binary against it on an isolated pgvector DB with 30 records and real River jobs:

| Config | Requests per pod |
|---|---|
| `MAX_CONCURRENT=1`, keep-alives on | **30 / 0 / 0** |
| `MAX_CONCURRENT=2`, keep-alives on | **16 / 14 / 0** |
| `MAX_CONCURRENT=5`, keep-alives on | 8 / 11 / 11 |
| `MAX_CONCURRENT=1`, setting **enabled** | **10 / 10 / 10** ✅ |

So the mechanism is exactly Go's `MaxIdleConnsPerHost` default of 2 — the worker reaches at most two TEI pods no matter how many replicas exist. The fix does what it says on the tin.

Nothing blocking from me. Two things I think are worth a look before merge, then some smaller stuff.

---

### 1. `cmd/backfill-embeddings` doesn't pick the setting up

`cmd/backfill-embeddings/main.go:117` builds its own `EmbeddingClientConfig` and omits `HTTPDisableKeepAlives`, so it silently keeps keep-alives even when the operator has set the env var. Since it reads the same `cfg.Embedding` block, that's a surprising place to diverge.

It's also the worst case for pinning: `main.go:229` registers `EmbeddingsQueueName: {MaxWorkers: 1}` (strictly sequential) and it never wraps in `NewBatchingEmbeddingClient` — so a full-corpus backfill sends every single embedding down one connection to one pod. The PR body explains why the API path is deliberately excluded; the backfill isn't mentioned, so I'm reading this as an oversight rather than a decision — but happy to be told otherwise.

### 2. Will this actually move the throughput number?

I think this belongs to ENG-2561 ("creates 1 embedding / second"), though nothing links the two — could we get the ticket on the branch/PR title so it's traceable?

My worry is that on the chart's own defaults this won't help much yet:

- `charts/hub/values.yaml` defaults `embeddings.replicaCount: 1`, with autoscaling disabled and `maxReplicas: 2`. At ≤2 replicas the existing transport already reaches every pod (see the `MAX_CONCURRENT=5` row above), so the setting buys nothing there and only adds a handshake per request.
- Meanwhile the chart has **no** `EMBEDDING_BATCH_*` key at all, so production runs `EMBEDDING_BATCH_SIZE=1` — micro-batching off, one HTTP request per embedding — against the CPU-only TEI image. That feels like the more likely source of ~1/sec.

Not an argument against this change, it's a real fix for a real thing. Just wondering whether the pinning was observed in the cluster, and at what replica count — otherwise batching + replicas might be the bigger lever and this lands as groundwork.

### 3. The setting is never logged in the default configuration

`cmd/worker/app.go:179` puts `http_disable_keep_alives` inside the `if enabled` branch for batching, and `NewBatchingEmbeddingClient` bails when `BatchSize <= 1` — which is the default. I booted the worker with `EMBEDDING_HTTP_DISABLE_KEEP_ALIVES=true` and the default batch size and got **zero** log lines mentioning it, while the transport change was clearly active (the 10/10/10 row).

For an ops-only knob whose whole point is diagnosing distribution, that's the one line you want at startup — could it move next to the provider/base-URL line so it's unconditional?

### 4. No chart surface

The chart lives in this repo and deploys the very replicas we're spreading across, and it exposes the sibling knobs first-class (`embeddings.maxConcurrent`, `embeddings.normalize`). This one has no `values.yaml` key, so operators have to hand-roll `worker.extraEnv`. It *does* get through — I checked `hub.embeddingEnvManaged` and it filters only the six managed keys — so not a blocker, just inconsistent.

### 5. Smaller stuff

- `HTTPDisableKeepAlives` is openai-only (only `openAIEmbeddingFactory` reads it), and `enrichment_client_factory.go:112` already *hard-errors* when an openai-only knob is set on another provider. This one is silently ignored for google/gemini. The `.env.example` wording covers it now, so this is opinion rather than a defect.
- `internal/openai/client.go:94` — the unchecked `.(*http.Transport)` would panic at construction if anything ever wrapped `http.DefaultTransport` (nothing in-tree does; otelhttp is server-side only here). It's the canonical Go idiom so I'd probably leave it, just flagging.
- `embedding_batcher_test.go:388` — the 1s `deadline` is the same value as the batcher's `MaxWait`. Batches of 2 form immediately so it never actually depends on the timer, but that's a thin margin on a loaded runner. `2*time.Second` costs nothing.
- Worth noting `TestBatchingEmbeddingClientLimitsConcurrentProviderRequests` isn't coverage for this diff (no batcher production code changed) — it's a bonus regression test for the existing `MaxInFlight`. Good to have, I did check it can fail: widening the semaphore to `MaxInFlight*2` trips `provider exceeded MaxInFlight=12`.

---

### Security pass — clean

Checked the things a hand-rolled transport usually gets wrong, rather than assuming:

- **TLS verification intact.** The clone keeps `TLSClientConfig: nil`, so system roots. Pointed the client at a self-signed TLS server and it correctly failed with `x509: certificate signed by unknown authority`.
- **Egress proxy still honoured.** `Clone()` preserves `Proxy: ProxyFromEnvironment` — with `HTTP_PROXY` set the request went through a stub proxy.
- **No timeout regression.** The SDK default is `http.DefaultClient` (zero `Timeout`) and the replacement is `&http.Client{Transport: …}` (also zero), so identical. `WithMaxRetries(0)` untouched, and `FeedbackEmbeddingWorker.Timeout` still bounds the job, so the extra dials can't wedge a slot.
- **Not a silent no-op over HTTP/2** — this was my main worry. Against an h2 TLS server, `DisableKeepAlives` makes Go's h2 transport use single-use connections (4 requests → 4 TCP connections), so it holds there too, not just HTTP/1.1.
- The cost worth stating next to the benefit: a handshake and an ephemeral port per request means more TIME_WAIT on the worker and more handshake CPU on TEI. Default-off so I'm not worried, and I measured the overhead as noise (2.925s vs 2.926s for 30 sequential jobs over loopback; in-cluster a sub-ms handshake against ~100ms calls is under 1%).

Also ran 20× `-race` on both new tests (no flakes), the full unit suite, `go vet` and `golangci-lint` on the touched packages — all clean, and CI is green across all 11 checks.

**Comment by BhagyaAmarasinghe** [comment]
Addressed the review in `7884a26`:

1. `backfill-embeddings` now propagates `HTTPDisableKeepAlives` into its client config. One important clarification: the command is producer-only. It registers the embedding job kind because River requires it, but never starts the River client; `hub-worker` executes the requests, so the command itself cannot pin a sequential provider connection.
2. The PR title and description now link ENG-2561 and record the actual staging evidence: six Ready TEI replicas, requests reaching only two pods, max in-flight 3, and 500 records in 382 seconds (1.31/s). The description also links formbricks/formbricks#9001, which carries the worker-only chart surface and the `48 / 8 / 100 / 12` staging tuning used for the guarded rerun.
3. `http_disable_keep_alives` is now logged whenever the embedding worker is configured, including the default non-batching path.
4. The production Formbricks chart surface is in formbricks/formbricks#9001. I did not expand this PR into the standalone Hub chart; that chart can still opt in through `worker.extraEnv`, and it is not the chart used by this staged rollout.
5. The concurrency test deadline is now 2 seconds. I kept the canonical cloned `http.DefaultTransport` assertion and the documented OpenAI-only behavior unchanged; tightening validation for non-OpenAI providers would be a separate compatibility change.

Validation on the new head:

- `go test ./cmd/backfill-embeddings ./cmd/worker ./internal/service`
- `go test -race ./internal/service`
- `make build-backfill-embeddings`
- `make build`
- `make test-unit`
- `make GOLANGCI_LINT="$(go env GOPATH)/bin/golangci-lint" lint`

All pass locally. Exact-head GitHub checks are running.

**Review by xernobyl — APPROVED** [review]
Approving at `7884a26d`. All five points from the last round are properly addressed, and one of them you were right to push back on.

## Verified

**The backfill clarification is correct, and it corrects my reasoning rather than just the code.** My worst case was "a full-corpus backfill sends every embedding down one connection to one pod". I checked: `riverClient` is constructed but `.Start()` is never called anywhere in `cmd/backfill-embeddings` — `cmd/worker/app.go:387` is the only place a River client starts — and `embeddingClient` is only handed to a worker that never runs. So the command genuinely cannot issue an embedding request. Propagating the field is still the right call for config consistency, but the pinning scenario I attached to it does not exist. Thanks for saying so instead of just changing the code.

**The log fix works.** I booted the worker binary against an isolated pgvector database rather than reading the diff:

| Configuration | Result |
| --- | --- |
| env var absent | `embedding worker configured provider=openai http_disable_keep_alives=false` |
| set to `true`, default batch size | logged once |
| set to `true`, batching on | logged once, no duplicate |

That was the case that produced zero lines before, so the ops knob is now visible in exactly the configuration that matters.

One thing I ran into and want to record as *not* a problem: setting the variable to an empty string fails the worker at boot with `strconv.ParseBool: parsing "": invalid syntax`. That is the pre-existing cleanenv behaviour shared with the sibling `Normalize` bool, and formbricks/formbricks#9001 always renders a quoted `"true"`/`"false"` (defaulting to `"false"`, with CI asserting both), so it is unreachable through the chart. Noting it only so nobody re-discovers it and files it as a regression here.

## Checks I ran

`go build` including both binaries, `go vet`, the full unit suite, `golangci-lint` (0 issues), both named regression tests, `-race -count=15` on each of them (no flakes), `-race` across config/openai/service, and the integration suite.

The branch is two commits behind `main` and those commits touch four of the same files — `.env.example`, `cmd/worker/app.go`, `internal/config/config.go`, `internal/config/config_test.go` — so I merged `origin/main` locally and re-ran build, vet, unit, lint and integration rather than trusting `MERGEABLE`. Clean, and both fixes survive the merge.

I also re-ran the mutation check on the new head: removing `transport.DisableKeepAlives = true` turns `opens_a_connection_per_request_when_disabled` red, so the guard is still real after the refactor.

## Security pass — clean

Re-verified rather than carried over from the last round, since a hand-rolled transport is worth checking twice:

- **TLS verification enforced on both paths.** Against a self-signed TLS server the client fails with an x509 error whether keep-alives are on or off.
- **The clone preserves `Proxy` and `ForceAttemptHTTP2`**, and `TLSClientConfig` carries no `InsecureSkipVerify`. My end-to-end proxy probe was inconclusive — Go caches `HTTP_PROXY` behind a `sync.Once`, so `t.Setenv` is ignored once another test in the package has run — so I asserted on the cloned fields directly instead of trusting a negative result.
- **`http.DefaultTransport` is not mutated.** Worth stating explicitly: the naive version of this change sets `DisableKeepAlives` on the shared default and silently affects every other HTTP client in the process. Cloning first avoids that, and I confirmed the base transport is untouched afterwards.
- **No secrets in the new log line** — provider name and a boolean. The `ProviderAPIKey` and `BaseURL` lines in the diff are config-struct assignments, not log fields.

## Two notes, neither blocking

**1. Neither behavioural fix is covered by a test.** `7884a26d` touched three files and only the deadline change touched a test. There are no test files in `cmd/worker` or `cmd/backfill-embeddings` at all — `cmd/api` is the only `cmd` package with any — so both the propagation and the log line could regress silently. Most pointed for the log line, since its absence was the original finding and it exists specifically to diagnose this setting in production. Worth a follow-up rather than a change here.

**2. CI has not run the tests for this head yet.** `Tests` and `API Contract Tests` are still queued, `Code Quality` is the only green check, and `preview` reads as failed but is actually *cancelled* inside `Build SDKs for pull request` — which `main` just reworked in `d8fb94e` when the Hub docs moved off Stainless, and this branch predates that. My local runs cover the same ground, but I would let those two land before merging rather than treating this approval as check-complete.

Approving the code on that basis — the change is focused, the regression test genuinely guards it, and the throughput reasoning in the description now matches the evidence.
